### PR TITLE
transport: Remove buffer copies while writing HTTP/2 Data frames

### PR DESCRIPTION
**Pull request by arjan-bal**
This PR removes 2 buffer copies while writing data frames to the underlying net.Conn: one [within gRPC](https://github.com/grpc/grpc-go/blob/58d4b2b1492dbcfdf26daa7ed93830ebb871faf1/internal/transport/controlbuf.go#L1009-L1022) and the other [in the framer](https://cs.opensource.google/go/x/net/+/master:http2/frame.go;l=743;drc=6e243da531559f8c99439dabc7647dec07191f9b). Care is taken to avoid any extra heap allocations which can affect performance for smaller payloads.

A [CL](https://go-review.git.corp.google.com/c/net/+/711620) is out for review which allows using the framer to write frame headers. This PR duplicates the header writing code as a temporary workaround. This PR will be merged only after the CL is merged.

### Small payloads
Performance for small payloads increases slightly due to the reduction of a `deferred` statement.
```
$ go run benchmark/benchmain/main.go -benchtime=60s -workloads=unary \
   -compression=off -maxConcurrentCalls=120 -trace=off \
   -reqSizeBytes=100 -respSizeBytes=100 -networkMode=Local -resultFile="${RUN_NAME}"

$ go run benchmark/benchresult/main.go unary-before unary-after
               Title       Before        After Percentage
            TotalOps      7600878      7653522     0.69%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op     10007.07     10000.89    -0.07%
           Allocs/op       146.93       146.91     0.00%
             ReqT/op 101345040.00 102046960.00     0.69%
            RespT/op 101345040.00 102046960.00     0.69%
            50th-Lat    833.724µs    830.041µs    -0.44%
            90th-Lat   1.281969ms   1.275336ms    -0.52%
            99th-Lat   2.403961ms   2.360606ms    -1.80%
             Avg-Lat    946.123µs    939.734µs    -0.68%
           GoVersion     go1.24.8     go1.24.8
         GrpcVersion   1.77.0-dev   1.77.0-dev
```

### Large payloads
Local benchmarks show a ~5-10% regression with 1 MB payloads on my dev machine. The profiles show increased time spent in the copy operation [inside the buffered writer](https://github.com/grpc/grpc-go/blob/58d4b2b1492dbcfdf26daa7ed93830ebb871faf1/internal/transport/http_util.go#L334). Counterintuitively, copying the grpc header and message data into a larger buffer increased the performance by 4% (compared to master).

To validate this behaviour (extra copy increasing performance) I ran [the k8s benchmark for 1MB payloads](https://github.com/grpc/grpc/blob/65c9be86830b0e423dd970c066c69a06a9240298/tools/run_tests/performance/scenario_config.py#L291-L305) and 100 concurrent streams which showed ~5% increase in QPS without the copies across multiple runs. Adding a copy reduced the performance.

Load test config file:  [loadtest.yaml](https://github.com/user-attachments/files/23055312/loadtest.yaml)

```
# 30 core client and server
Before
QPS: 498.284 (16.6095/server core)
Latencies (50/90/95/99/99.9%-ile): 233256/275972/281250/291803/298533 us
Server system time: 93.0164
Server user time:   142.533
Client system time: 97.2688
Client user time:   144.542

After
QPS: 526.776 (17.5592/server core)
Latencies (50/90/95/99/99.9%-ile): 211010/263189/270969/280656/288828 us
Server system time: 96.5959
Server user time:   147.668
Client system time: 101.973
Client user time:   150.234

# 8 core client and server
Before
QPS: 291.049 (36.3811/server core)
Latencies (50/90/95/99/99.9%-ile): 294552/685822/903554/1.48399e+06/1.50757e+06 us
Server system time: 49.0355
Server user time:   87.1783
Client system time: 60.1945
Client user time:   103.633

After
QPS: 334.119 (41.7649/server core)
Latencies (50/90/95/99/99.9%-ile): 279395/518849/706327/1.09273e+06/1.11629e+06 us
Server system time: 69.3136
Server user time:   102.549
Client system time: 80.9804
Client user time:   107.103
```

RELEASE NOTES:
* transport: Avoid two buffer copies when writing data.

### CODE REVIEW

File: mem/buffer_slice.go
Comment (easwars) [inline mem/buffer_slice.go]: While you are here, maybe you can remove mentions of this one additional convenience method. Looks like there are going to be multiple convenience methods going forward.

File: mem/buffer_slice.go [hunk 1]
```diff
@@ -140,6 +140,17 @@ type Reader interface {
 	// Reset frees the currently held buffer slice and starts reading from the
 	// provided slice. This allows reusing the reader object.
 	Reset(s BufferSlice)
+	// Peek returns up to the next n bytes from the reader's current position as
+	// a slice of byte slices.
+	// The reader's position is not advanced.
+	// Next appends results to the provided res slice and returns the updated
+	// slice.
```
Comment (easwars) [inline mem/buffer_slice.go]: Nit: This line about Next seems out of place. Maybe delete?

File: mem/buffer_slice.go [hunk 2]
```diff
@@ -140,6 +140,17 @@ type Reader interface {
 	// Reset frees the currently held buffer slice and starts reading from the
 	// provided slice. This allows reusing the reader object.
 	Reset(s BufferSlice)
+	// Peek returns up to the next n bytes from the reader's current position as
+	// a slice of byte slices.
```
Comment (easwars) [inline mem/buffer_slice.go]: Does it make sense to return the actual number of bytes returned in `res`? Most APIs to read data do that.

File: mem/buffer_slice.go [hunk 3]
```diff
@@ -140,6 +140,17 @@ type Reader interface {
...
+	// The reader's position is not advanced.
+	// Next appends results to the provided res slice and returns the updated
+	// slice.
+	// The returned subslices are views into the underlying buffers and are only
+	// valid until the reader is advanced past the corresponding buffer.
+	Peek(n int, res [][]byte) [][]byte
```
Comment (easwars) [inline mem/buffer_slice.go]: It is not clear to me as to why the result is both a parameter and a return value in this method.

File: mem/buffer_slice_test.go [hunk 4]
```diff
@@ -484,3 +484,143 @@ func (t *testPool) Put(buf *[]byte) {
...
+	}{
+		{
+			name: "empty",
+			operations: func(t *testing.T, c mem.Reader) {
+				if c.Remaining() != 0 {
+					t.Errorf("Remaining() = %v, want 0", c.Remaining())
+				}
+				res := c.Peek(1, nil)
+				if len(res) != 0 {
+					t.Errorf("Peek() got %v, want empty", res)
```
Comment (easwars) [inline mem/buffer_slice_test.go]: Nit: `Peek() got %v slices, want empty`?

File: mem/buffer_slice_test.go [hunk 5]
```diff
@@ -484,3 +484,143 @@ func (t *testPool) Put(buf *[]byte) {
 	}
 	delete(t.allocated, buf)
 }
+
+func (s) TestBufferSlice_Iteration(t *testing.T) {
+	tests := []struct {
+		name       string
+		buffers    [][]byte
+		operations func(t *testing.T, c mem.Reader)
```
Comment (easwars) [inline mem/buffer_slice_test.go]: It does feel like many of checks in many of the operations should result in calls to `t.Fatal` instead of `t.Error` since it doesn't make sense to carry on (for example, when you start with an empty buffer and `Remaining` doesn't return `0`).

And if we start calling `t.Fatal` at different places, we also probably need to defer the call to the `Close` method of the reader in the main test logic.

File: mem/buffer_slice_test.go [hunk 6]
```diff
@@ -484,3 +484,143 @@ func (t *testPool) Put(buf *[]byte) {
...
+	}{
+		{
+			name: "empty",
+			operations: func(t *testing.T, c mem.Reader) {
```
Comment (easwars) [inline mem/buffer_slice_test.go]: Curious: why is the reader called `c` here and in the main test logic?

File: internal/transport/http_util.go [hunk 7]
```diff
@@ -389,8 +389,9 @@ func toIOError(err error) error {
 }
 
 type framer struct {
-	writer *bufWriter
-	fr     *http2.Framer
+	writer    *bufWriter
+	fr        *http2.Framer
+	headerBuf []byte // cached slice for framer headers to reduce heap allocs.
```
Comment (easwars) [inline internal/transport/http_util.go:403]: Can this be an array instead? `[9]byte`?

File: internal/transport/http_util.go [hunk 8]
```diff
@@ -422,6 +423,45 @@ func newFramer(conn net.Conn, writeBufferSize, readBufferSize int, sharedWriteBu
...
+	f.headerBuf = append(f.headerBuf[:0],
+		byte(length>>16),
+		byte(length>>8),
+		byte(length),
+		byte(http2.FrameData),
+		byte(flags),
+		byte(streamID>>24),
+		byte(streamID>>16),
+		byte(streamID>>8),
+		byte(streamID))
```
Comment (easwars) [inline internal/transport/http_util.go:470]: And if we make the `headerBuf` an array, does it make sense to directly set entries in the array (using indices) instead of using `append`?

File: internal/transport/http_util.go [hunk 9]
```diff
@@ -422,6 +423,45 @@ func newFramer(conn net.Conn, writeBufferSize, readBufferSize int, sharedWriteBu
...
+		return err
+	}
+	for _, d := range data {
+		if len(d) == 0 {
+			continue
+		}
+		_, err := f.writer.Write(d)
+		if err != nil {
+			return err
+		}
```
Comment (easwars) [inline internal/transport/http_util.go]: Nit: assignment and conditional on the same line?

File: internal/transport/http_util.go (hunk 7 above)
Comment (arjan-bal) [reply internal/transport/http_util.go:403]: Yes, we can, but I see the following disadvantages:

* The buffered [writer accepts a slice](https://github.com/grpc/grpc-go/blob/254ab1095e9f4179cebd36517bfb7e61b623e509/internal/transport/http_util.go#L320), so we would need to slice the array (arr[:]).
* The slice would be passed to an io.Writer interface, so the compiler's escape analysis will likely send the array to the heap, losing the benefit of stack allocation.
* Arrays are value types, so passing them as function params without slicing would result in copies of the entire array.

Due to the above, I don't think using arrays would help here.

File: internal/transport/http_util.go (hunk 8 above)
Comment (arjan-bal) [reply internal/transport/http_util.go:470]: Responded to the argument for using slice above. We can use indices here, but using the append syntax achieves the same result with fewer characters.

File: internal/transport/http_util.go (hunk 9 above)
Comment (arjan-bal) [reply internal/transport/http_util.go]: Changed.

File: mem/buffer_slice.go (hunk 2 above)
Comment (arjan-bal) [reply mem/buffer_slice.go]: I was using the `bufio.Reader` type as a reference. It has a [Peek](https://pkg.go.dev/bufio#Reader.Peek) method that returns the resulting slice and an error. The number of bytes read can be derived from the length of the returned slice. I've updated the `Peek` method here to also return an error, even though it's not expected to fail in gRPC because of the way the requested size is calculated before the call.

This length is expected to be the same as the number of bytes requested. However, if the returned error is non-nil, the slice will be shorter.

Using `Reader.Peek()` and `Reader.Discard()` is a common pattern to avoid copies, as seen here: [https://www.ianlewis.org/en/zero-copy-readers-in-go#bufioreader](https://www.google.com/search?q=https.www.ianlewis.org/en/zero-copy-readers-in-go%23bufioreader). The APIs here are like a 2D version of `bufio.Reader`'s

File: mem/buffer_slice.go (hunk 1 above)
Comment (arjan-bal) [reply mem/buffer_slice.go]: This method was called `Next` earlier before I discovered that bufio had similar functionality. I've updated the godoc now.

File: mem/buffer_slice.go (hunk 3 above)
Comment (arjan-bal) [reply mem/buffer_slice.go]: This is a common technique to maximize buffer re-use and minimize allocations. If the passed destination buffer has insufficient capacity, a larger destination will be allocated and returned. This is similar to the built-in append function: `dest = append(dest, 1) // accepts and returns the destination`.

I found this pattern when working with ALTS code that called implementations of the AED interface: https://pkg.go.dev/crypto/cipher#AEAD

I've added a line in the godoc mentioning this.

File: mem/buffer_slice_test.go (hunk 6 above)
Comment (arjan-bal) [reply mem/buffer_slice_test.go]: Renamed to `r`. Before discovering the similar functionality in `bufio.Reader`, I had a new type called `Cursor`, so these variables were named `c`.

File: mem/buffer_slice_test.go (hunk 5 above)
Comment (arjan-bal) [reply mem/buffer_slice_test.go]: I wanted to have good coverage of all the edge cases so I asked Gemini to create tests before modifying them. The `t.Errorf` are an artifact of that. I've replaced them with `Fatalf` now.

The `defer reader.Close()` call is present.

File: mem/buffer_slice_test.go (hunk 4 above)
Comment (arjan-bal) [reply mem/buffer_slice_test.go]: Fixed.

File: mem/buffer_slice.go
Comment (arjan-bal) [reply mem/buffer_slice.go]: Updated the godoc to make is less specific.

**Comment by easwars** [comment]
@dfawley : Again moving to your plate if you feel like having a second look.

File: internal/transport/controlbuf.go [hunk 10]
```diff
@@ -999,25 +1000,18 @@ func (l *loopyWriter) processData() (bool, error) {
...
-			// Note that this is only supposed to be nil in tests. Otherwise, stream is
-			// always initialized with a BufferPool.
-			pool = mem.DefaultBufferPool()
+	l.writeBuf = l.writeBuf[:0]
+	if hSize > 0 {
+		l.writeBuf = append(l.writeBuf, dataItem.h[:hSize])
+	}
+	if dSize > 0 {
+		var err error
+		l.writeBuf, err = reader.Peek(dSize, l.writeBuf)
```
Comment (dfawley) [inline internal/transport/controlbuf.go:1020]: It seems like this buffer can only grow and never shrinks.

1. What happens if a slice holds a pointer to a huge amount of data?  I believe it isn't possible to free it, but am not certain.  E.g.

```go
l.writeBuf = [][]byte{nil, nil, nil, nil, nil, nil, make([]byte, 10GB)}
l.writeBuf = l.writeBuf[:0]
```

2. What happens if `cap(l.writeBuf)` grows to a large value and then we never need it to be that large ever again?

I think we need to have some way to scale this buffer back down.

File: internal/transport/http_util.go [hunk 11]
```diff
@@ -422,6 +423,44 @@ func newFramer(conn net.Conn, writeBufferSize, readBufferSize int, sharedWriteBu
...
+		byte(streamID>>16),
+		byte(streamID>>8),
+		byte(streamID))
+	if _, err := f.writer.Write(f.headerBuf); err != nil {
+		return err
+	}
+	for _, d := range data {
+		if len(d) == 0 {
+			continue
+		}
```
Comment (dfawley) [inline internal/transport/http_util.go]: Would this be a bug if it were zero?  I would have expected it to be.

If it is, then we should delete it.  `Write` should handle a zero-length buffer as a nop already anyway.

Comment (arjan-bal) [reply internal/transport/http_util.go]: Removed. There should not be any empty buffers in the list, since `Peek()` filters them out. This was an artifact from the time I spent root-causing unexpected behavior on the local benchmarks with large payloads

File: internal/transport/controlbuf.go (hunk 10 above)
Comment (arjan-bal) [reply internal/transport/controlbuf.go:1020]: For point 1, I've updated the code to clear the buffer after calling `Write`. This releases references to all the slices and allows them to be GCed.

With respect to point 2, I've now set a limit of 64 on the buffer's length. If a buffer is longer than that, it's immediately freed after use instead of being cached.

Background on the 64-element limit: The `BufferSlice` from the proto codec is 1 element. With a potential gRPC header, the length is almost always 2. While custom codecs might produce larger slices, 64 is a generous limit that covers common cases without caching excessive memory.

This change also mitigates a worst-case memory scenario. Since `Peek()` filters empty slices, a 16KB http2 Data frame (the max size) could theoretically be split into 16K (16,384) distinct 1-byte slices. In that case, the memory overhead for the slice *headers* alone would be `24 bytes * 16 * 1024` (approx. 393KB), with the 64 size limit, the max held memory is approx 1.5KB. Also note that the framer already [has a data buffer that grows up to 16KB](https://cs.opensource.google/go/x/net/+/refs/tags/v0.46.0:http2/frame.go;l=715;drc=10342476f5c8587b46c71128962c71d0c3cd6904), and after this change, that buffer should no longer be used for Data frames.

File: internal/transport/controlbuf.go [hunk 12]
```diff
@@ -1010,6 +1021,8 @@ func (l *loopyWriter) processData() (bool, error) {
 		if err != nil {
 			// This must never happen since the reader must have at least dSize
 			// bytes.
+			clear(l.writeBuf)
+			l.writeBuf = nil
```
Comment (dfawley) [inline internal/transport/controlbuf.go]: If this is impossible then:

1. `logger.Error` seems like a good idea, unless the caller already does that with what we return..
2. We probably don't need to bother with the clear/nil (and surely don't want to do both?)?

**Review by dfawley — APPROVED** [review]

File: internal/transport/controlbuf.go (hunk 12 above)
Comment (arjan-bal) [reply internal/transport/controlbuf.go]: Added an error log and removed the buffer resetting.
